File: java/bitcoin/Test_Abstract.java
import java.security.SecureRandom;

public abstract class Test_Abstract implements Runnable {

  public abstract void run();

  private static final SecureRandom _random = new SecureRandom();

  protected static byte[] getRandomBytes(int n)
  {
    byte[] bytes = new byte[n];
    _random.nextBytes(bytes);
    return bytes;
  }

  protected static void logMessage(String message)
  {
    System.err.println(message);
  }

  protected static void checkEquals(Object left, Object right, String msg)
  {
    if(left != null)
    {
      if(!left.equals(right))
      {
        logMessage(msg + ": checkEquals failure");
        logMessage("left = " + left);
        logMessage("right = " + right);
        System.exit(1);
      }
      // in case equals override is not symmetric
      if(!right.equals(left))
      {
        logMessage(msg + ": checkEquals failure");
        logMessage("override of equals method is not symmetric");
        logMessage("left.equals(right) is true while right.equals(left) is false");
        System.exit(1);
      }
          
    }
    else
    {
      if(right != null)
      {
        logMessage(msg + ": checkEquals failure");
        logMessage("left is null, but right is not");
        System.exit(1);
      }
    }
  }

  protected static void checkNotNull(Object obj, String msg)
  {
    if(obj == null)
    {
      logMessage(msg + ": checkNotNull failure");
      System.exit(1);
    }
  }

  protected static void checkCondition(boolean cond, String msg)
  {
    if(!cond)
    {
      logMessage(msg + ": checkCondition failure");
      System.exit(1);
    }
  }

  protected static void checkException(Runnable callbk, String name, String msg)
  {
    try
    {
      callbk.run();
      logMessage(msg + ": checkException failure: no exception detected");
      System.exit(1);
    }
    catch(Exception e)
    {
      if(!e.getClass().getName().equals("java.lang." + name))
      {
        logMessage(msg + ": checkException failure: wrong exception type");
        logMessage("Expected: java.lang." + name);
        logMessage("Actual: " + e.getClass().getName());
        System.exit(1);
      }
    }
  }
}

File: c#/bitcoin/Test_Abstract.cs
using System;
using System.Security.Cryptography;
using Random = System.Security.Cryptography.RandomNumberGenerator;

public abstract class Test_Abstract
{
  public abstract void Run();

  private static readonly Random _random = new RNGCryptoServiceProvider();

  protected delegate void Runnable();

  protected static byte[] GetRandomBytes(int n)
  {
    byte[] bytes = new byte[n];

    _random.GetBytes(bytes);

    return bytes;
  }

  protected static void LogMessage(string message)
  {
    Console.Error.WriteLine(message);
  }


  protected static void CheckEquals(object left, object right, string msg)
  {
    if(left != null)
    {
      if(!left.Equals(right))
      {
        LogMessage(msg + ": CheckEquals failure");
        LogMessage("left = " + left);
        LogMessage("right = " + right);
        Environment.Exit(1);
      }
    }
    else
    {
      if(right != null)
      {
        LogMessage(msg + ": CheckEquals failure");
        LogMessage("left is null, but right is not");
        Environment.Exit(1);  // java System.exit(1)
      }
    }
  }

  protected static void CheckNotNull(object obj, string msg)
  {
    if(obj == null)
    {
      LogMessage(msg + ": CheckNoNull failure");
      Environment.Exit(1);
    }
  }

  protected static void CheckCondition(bool cond, string msg)
  {
    if(!cond)
    {
      LogMessage(msg + ": CheckCondition failure");
      Environment.Exit(1);
    }
  }

  protected static void CheckException(Runnable callbk, string name, string msg)
  {
    try
    {
      callbk();
      LogMessage(msg + ": CheckException failure: no exception detected");
      Environment.Exit(1);
    }
    catch(Exception e)
    {
      if(!e.GetType().Name.Equals(name))
      {
        LogMessage(msg + ": CheckException failure: wrong exception type");
        LogMessage("Expected: " + name);
        LogMessage("Actual: " + e.GetType().Name);
        Environment.Exit(1);
      }
    }
  }

  protected static void CheckArrayEquals(byte[] a, byte[] b, string msg)
  {
    if(a.Length != b.Length)
    {
      LogMessage(msg + ": CheckArrayEquals failure");
      LogMessage("left has length" + a.Length);
      LogMessage("right has length {0}" + b.Length);
      Environment.Exit(1);
    }

    for(int i = 0; i < a.Length; ++i)
    {
      if(a[i] != b[i])
      {
        LogMessage(msg + ":CheckArrayEquals failure at index i = " + i);
        LogMessage("left = " + a[i]);
        LogMessage("right = " + b[i]);
      }
    }
  }
}
